File: 06-linked_list/00287-Find_The_Duplicate_Number/00287-find_the_duplicate_number.cpp
// 287: Find the Duplicate Number
// https://leetcode.com/problems/find-the-duplicate-number/


#include <iostream>
#include <vector>

using namespace std;

class Solution {
public:
    // SOLUTION
    int findDuplicate(vector<int>& nums) {        
        int slow = nums[0], fast = nums[nums[0]];
        
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[nums[fast]];
        }
        
        slow = 0;
        while (slow != fast) {
            slow = nums[slow];
            fast = nums[fast];
        }

        return slow;
    }
};



int main() {
    Solution o;

    // INPUT
    vector<int> nums = {1,3,4,2,2};

    // OUTPUT
    auto result = o.findDuplicate(nums);
    cout << result << endl;

    return 0;
}

File: 06-linked_list/00287-Find_The_Duplicate_Number/00287-find_the_duplicate_number.py
# 287: Find the Duplicate Number
# https://leetcode.com/problems/find-the-duplicate-number/


class Solution:
    # SOLUTION
    def findDuplicate(self, nums: list[int]) -> int:
        slow, fast = nums[0], nums[nums[0]]
        
        while slow != fast:
            slow = nums[slow]
            fast = nums[nums[fast]]
        
        slow = 0
        while slow != fast:
            slow = nums[slow]
            fast = nums[fast]

        return slow


if __name__ == "__main__":
    o = Solution()

    # INPUT
    nums: list[int] = [1,3,4,2,2]
    
    # OUTPUT
    result = o.findDuplicate(nums)
    print(result)